### DISCUSSION TITLE
[Still not an optimal experience] C# 9.0 provides generics that supports both nullable classes and structs, but it only does it 50%

### DISCUSSION THREAD

**Discussion by jannickj** [dotnet/csharplang#4118] [General]
Example assume I want to provide a nullable wrapper for dictionary's TryGetValue
```csharp
public static class DictionaryExt
{

	public static TVal? TryGet<TKey, TVal>(this Dictionary<TKey, TVal> dic, TKey key)
		where TKey : notnull
	{
		if (dic.TryGetValue(key, out var val))
		{
			return val;
		}
		return null;
	}
```
this gives you `Cannot convert null to type parameter 'TVal' because it could be a non-nullable value type. Consider using 'default(TVal)' instead.`
Now this is terrible advice if you actually do this i.e. `null` -> `default` then
```csharp
var dict = new Dictionary<int, int>();
var val = dict.TryGet(0);
> val.GetType() == typeof(int) && val == 0
true
```

instead what you actually have to do (like you did in c# 8.0)
```csharp
public static class DictionaryExtStruct
{
	public static TVal? TryGet<TKey, TVal>(this Dictionary<TKey, TVal> dic, TKey key)
		where TKey : notnull
		where TVal : struct
	{
		if (dic.TryGetValue(key, out var val))
		{
			return val;
		}
		return null;
	}
}

public static class DictionaryExtClass
{
	public static TVal? TryGet<TKey, TVal>(this Dictionary<TKey, TVal> dic, TKey key)
		where TKey : notnull
		where TVal : class
	{
		if (dic.TryGetValue(key, out var val))
		{
			return val;
		}
		return null;
	}
}
```

All of it coming from the fact that `null` is not valid return type from something that returns `[type]?` which is obviously incorrect.

**Comment by Joe4evr** [comment]
This is intentional. You can't change a struct substitution to a `Nullable<T>` from an unconstrained generic parameter without CLR changes.

**Reply by jannickj** [reply]
I updatede the headline, however I think at the very least the suggestion should be updated as it makes it sound like it would work otherwise.
I was thinking that we could fix it in the compiler by just making it auto sugar it, however given that the `static class` needs to be different this would probably be more confusing. :disappointed: 

I guess this is something that could be fixed by creating a library using the new source generators. :pray:

**Reply by CyrusNajmabadi** [reply]
> however I think at the very least the suggestion should be updated as it makes it sound like it would work otherwise.

I think the suggestion is a good one.  'null' is not convertible to `T?`.  However, `default` is.  Given that it looks like the code is trying to get a default value here for that type, that's a suitable suggestion as a workaround.

Note: tihs is the point of suggestions.  The code is clearly in error, and it's normally the case then that the compiler has to guess as to what you mean.  This is a perfectly good guess here, so i think it's reasonable to offer it.

**Reply by jannickj** [reply]
@CyrusNajmabadi the point is the compiler will silently remove `?` from return type, and default of Nullable<int> is null which would be expected but default of int is 0.
That said maybe you're right and the problem is you don't get a compiler warning that the `?` is ignored.

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi the point is the compiler will silently remove ? from return type

yes.  It's an unconstrained generic. That's literally how it was designed to work :)   

> and the problem is you don't get a compiler warning that the ? is ignored.

It's not that it's ignored, it's that this is literally what this code means (which was very intentional).  This feature is exactly aobut allowing people to annotate the tons of APIs out there that have been shipping since C#2.0 and work in this exact fashion.  NRT is about annotating what the ecosystem already does.  Not about adding new functionality that has never been there before.

**Reply by jannickj** [reply]
> This feature is exactly aobut allowing people to annotate the tons of APIs out there that have been shipping since C#2.0 

Totally fair, that said it's still completely unexpected behavior (to me) that `T? func<T>()` resolves to `int func<int>()` and not `int? func<int>()`.
However I do understand why that is not possible to explain in the API, because it's not `T?` it's `class(T?) or struct(default(T))`, which once you specify `T: struct` allows you to use `struct(T?)`. 

It is totally a solvable problem just do like Rust by compiling it down to a Tagged Union. However as @Joe4evr said that would probably require a deeper redesign.

**Reply by Joe4evr** [reply]
> Totally fair, that said it's still completely unexpected behavior (to me) that `T? func<T>()` resolves to `int func<int>()` and not `int? func<int>()`

To note, I predicted *this exact reaction* back when the LDM was first considering changing `T?` to mean "nullable-only-if-nullable" when used on an unconstrained type parameter, and that it would've be better if it could've been left longer in case a better syntax option would come up later.

**Reply by alrz** [reply]
>  it's still completely unexpected behavior 

What would you expect? `T` is unconstrained so it can't really be `Nullable<T>`.

I find it only intuitive to say `T?` because that does make sense in the method declaration itself. 

It's only from caller's point of view that it might locally inferred as `int` which is actually the value `default(int)`,
that's why it's called `FirstOrDefault` not `FirstOrNull`.

I think non-defaultable structs (https://github.com/dotnet/csharplang/issues/146) will make this more relevant.

**Reply by jannickj** [reply]
> What would you expect? `T` is unconstrained so it can't really be `Nullable<T>`.

`T` can't be `Nullable<T>` but `T?` could easily be `Nullable<T>` The whole point of a type function (generics) is that the compiler can resolve the actual types as it pleases. If I write `int?` the compiler will happily resolve it as `Nullable<int>`. So I don't see why it can't be the same for generics (other than legacy).

> It's only from caller's point of view that it might locally inferred as `int` which is actually the value `default(int)`,
> that's why it's called `FirstOrDefault` not `FirstOrNull`.

`FirstOrDefault` is probably also the biggest antipattern function in the language, the first thing I do to any code base is create a `FirstOrNull` function. If Null exceptions is a billion dollar mistake, then I don't want to think about the dollar size of silent struct defaulting.

**Reply by alrz** [reply]
>So I don't see why it can't be the same for generics

`Nullable<T>` is constrained to structs only, that's why. Though it might be possible to use some kind of `Option<T>` type that works for unconstrained types but without runtime support it will suffer from an additional unnecessary boolean flag being carried around for reference types.

You may argue that we could postpone this feature until we have a runtime solution and I'd totally agree with that.

**Reply by jannickj** [reply]
> `Nullable<T>` is constrained to structs only, that's why. Though it might be possible to use some kind of `Option<T>` type that works for unconstrained types but without runtime support it will suffer from an additional unnecessary boolean flag being carried around for reference types.

No my point is that `if T is struct then T? -> Nullable<T>` and `if T is class then T? -> T` just like if I write `typeof(int?) == typeof(Nullable<int>)` and `typeof(object?) == typeof(object)`

**Reply by HaloFour** [reply]
The compiler doesn't resolve generics, the runtime does, and the runtime both doesn't understand NRTs nor does it support replacing the generic type in that manner.  There's not much the compiler can do here without runtime support, short of trying to emit two completely separate members.

**Reply by jannickj** [reply]
> The compiler doesn't resolve generics, the runtime does, and the runtime both doesn't understand NRTs nor does it support replacing the generic type in that manner. There's not much the compiler can do here without runtime support, short of trying to emit two completely separate members.

Fair enough I didn't actually know that :sweat_smile:, guess I learned something new today :)
